src/newsimplification/Simplifier.ts

```diff
@@ -0,0 +1,150 @@
+
+export class Simplifier {
+  static defaultOptions: SimplificationOptions = {
+    allowInheritance: false
+  }
+
+  private anonymCounter = 1;
+  private seenSchemas: Map<Schema, CommonModel> = new Map();
+  private iteratedModels: Record<string, CommonModel> = {};
+  
+  constructor(
+    readonly options: SimplificationOptions = Simplifier.defaultOptions,
+  ) {
+    this.options = { ...Simplifier.defaultOptions, ...options };
+  }
+
+  /**
+   * Simplifies a schema into instances of CommonModel. 
+   * Index 0 will always be the root schema CommonModel representation
+   * 
+   * @param schema to simplify
+   */
+  simplify(schema: Schema | boolean, splitModels = true): CommonModel[] {
+    const modelsToReturn = Object.values(this.iteratedModels);
+    if (typeof schema !== 'boolean' && this.seenSchemas.has(schema)) {
+      const cachedModel = this.seenSchemas.get(schema); 
+      if (cachedModel !== undefined) {
+        return [cachedModel];
+      }
+    }
+    //If it is a false validation schema return no common model
+    if (schema === false) {
+      return [];
+    }
+    const model = new CommonModel();
+    model.originalSchema = Schema.toSchema(schema);
+    if (schema === true) {
+      model.setType(['object', 'string', 'number', 'array', 'boolean', 'null', 'integer']);
+    }
+    if (typeof schema !== 'boolean') {
+      this.seenSchemas.set(schema, model);
+      this.simplifyModel(model, schema);
+    }
+    if (splitModels) {
+      this.ensureModelsAreSplit(model);
+    }
+    if (isModelObject(model)) {
+      this.iteratedModels[`${model.$id}`] = model;
+    }
+    return [model, ...modelsToReturn];
+  }
+
+  /**
+   * Function to simplify schema into a model.
+   * 
+   * @param model to simplify to simplify schema into 
+   * @param schema to simplify
+   */
+  private simplifyModel(model: CommonModel, schema: Schema) {
+    if (schema.type !== undefined) {
+      model.type = schema.type;
+    }
+
+    //All schemas of type object MUST have ids
+    if (model.type !== undefined && model.type.includes('object')) {
+      model.$id = simplifyName(schema) || `anonymSchema${this.anonymCounter++}`;
+    } else if (schema.$id !== undefined) {
+      model.$id = simplifyName(schema);
+    }
+
+    if (typeof schema !== 'boolean' && schema.required !== undefined) {
+      model.required = schema.required;
+    }
+
+    simplifyProperties(schema, model, this);
+    simplifyAllOf(schema, model, this);
+
+    this.combineSchemas(schema.oneOf, model, schema);
+    this.combineSchemas(schema.anyOf, model, schema);
+    this.combineSchemas(schema.then, model, schema);
+    this.combineSchemas(schema.else, model, schema);
+  }
+
+  /**
+   * Go through schema(s) and combine the simplified models together.
+   * 
+   * @param schema to go through
+   * @param currentModel the current output
+   */
+  combineSchemas(schema: (Schema | boolean) | (Schema | boolean)[] | undefined, currentModel: CommonModel, rootSchema: Schema) {
```

Comment:
Single Responsibility Principle! Instead of `| undefined` you should use `?`.

Reply:
Hehe 🙂 

I think this should be changed depending on the outcome of https://github.com/asyncapi/generator-model-sdk/pull/184#discussion_r632269074

Since we are using `| undefined` consistently I think we should keep it.

Reply:
Please do not use `| undefined` consistently. There is a slight difference between `?` and `| undefined`.
In objects `?` mean that key may not exists but `| undefined` mean that key must exists but its value can be undefined.

Reply:
Also you reduce codebase by 10 chars each time you replace `?` with `| undefined`.

Reply:
`?` would also force us to refactor the argument of the function and the way these functions are always called with potential `undefined` values, where the key will always exist. But I agree that in other cases it makes sense to use `?` as we are doing for properties.

Feel free to create an issue if you feel strongly about it and suggest we should change it 🙂

Reply:
@czlowiek488 But schema is in first argument, you cannot use optional statement here. Also schema should be used as first argument to function so `| undefined` is ok.